build.zig

```diff
@@ -1,42 +1,20 @@
 const std = @import("std");
 const Builder = std.build.Builder;
 const fmt = std.fmt;
-const printerr = std.debug.print;
 
 pub fn build(b: *Builder) void {
-    // error: array literal requires address-of operator to coerce to slice type '[][]const u8'
-    // var days: [][]const u8 = [_][]const u8{"day1", "day2"};
+    var buf = [_]u8{undefined} ** 100;
 
-    var days = [_][]const u8{"day1", "day2", "day3"};
-    var day_no: i32 = 1;
-    const allocator: *std.mem.Allocator = std.heap.page_allocator;
+    for ([_][]const u8{ "day1", "day2", "day3" }) |day, ix| {
+        var day_no = ix + 1;
 
-    for (days) |day| {
-        // Oops...
-        // var path_buf: [100]u8 = undefined;
-
-        var path_buf: []u8 = allocator.alloc(u8, 100) catch {
-            printerr("Error allocating memory for {}.", .{day});
-            return;
-        };
-        // Segmentation fault at address 0x7f066fb6b000
-        // defer allocator.free(path_buf);
-
-        var msg_buf: []u8 = allocator.alloc(u8, 100) catch {
-            printerr("Error allocating msg for {}.", .{day});
-            return;
-        };
-        // defer allocator.free(msg_buf);
-
-        const path_name = fmt.bufPrint(path_buf[0..], "{}/{}.zig", .{ day, day }) catch unreachable;
-        const day_exe = b.addExecutable(day, path_name);
+        const path_name = fmt.bufPrint(&buf, "{}/{}.zig", .{ day, day }) catch unreachable;
+        const day_exe = b.addExecutable(day, b.dupe(path_name));
         day_exe.setBuildMode(b.standardReleaseOptions());
 
         const day_cmd = day_exe.run();
-        const run_msg = fmt.bufPrint(msg_buf[0..], "Run AOC 2016 Day {} Program", .{ day_no }) catch unreachable;
-        const day_step = b.step(day, run_msg);
+        const run_msg = fmt.bufPrint(&buf, "Run AOC 2016 Day {} Program", .{day_no}) catch unreachable;
+        const day_step = b.step(day, b.dupe(run_msg));
```

Comment:
This is definitely a better solution, and I didn't know about it until this PR.

One question I have in general about `Builder` is... does the `Builder` "own" the buffers passed into it, and thus is guaranteed to free them? Or can it also accept const static buffers where `free`ing doesn't make sense?

If buffers aren't guaranteed to be `free`'d when a `Builder` is torn down<sup>1</sup>, I figured it might be possible to abuse `defer` to make the buffers go out of scope when the `Builder` does if the buffer was allocated, and the buffer is left alone if it _wasn't_ allocated :D.

1. Conditionally freeing could be possible if the buffer knows how to _destruct_ itself, rather than having the Builder call a `free` function to cause the deallocation. Global (and `const`) buffers would still be in scope- and thus wouldn't destruct themselves- when `Builder` does out of scope.

Reply:
`Builder` does not own the buffers passed into it, hence why we can just say `b.step("x", "y")` and not have it crash or be weird on us (i.e. const static buffers like you say). This is why we call [`b.dupe`](https://github.com/ziglang/zig/blob/6fb105fdd7798dc988de09a7b6709c5168355dfa/lib/std/build.zig#L293-L295) — it uses the `Builder`'s own allocator, and that _does_ leave it up to the (creator of the) `Builder` to free.

`Builder`'s own code uses `dupe` freely, and we can see that when a `Builder` is actually used and invoked, [it is created with an `ArenaAllocator`](https://github.com/ziglang/zig/blob/6fb105fdd7798dc988de09a7b6709c5168355dfa/lib/std/special/build_runner.zig#L22-L45), supporting this use pattern.

(I get the impression there's still much to be defined here, hence Zig still being at 0.6+, so a lot of the time you need to infer best practices from the stdlib and how it uses itself!)